connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.integration;
+
+import org.apache.kafka.connect.storage.StringConverter;
+import org.apache.kafka.connect.util.clusters.EmbeddedConnectCluster;
+import org.apache.kafka.test.IntegrationTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.kafka.connect.runtime.ConnectorConfig.CONNECTOR_CLASS_CONFIG;
+import static org.apache.kafka.connect.runtime.ConnectorConfig.TASKS_MAX_CONFIG;
+
+/**
+ * A simple integration test which brings up Connect, Kafka and ZooKeeper and produces messages
+ * into multiple topic-partitions, and sinks them with multiple tasks of a {@link MonitorableSinkConnector}.
+ */
+@Category(IntegrationTest.class)
+public class ConnectIntegrationTest {
+
+    private static final Logger log = LoggerFactory.getLogger(ConnectIntegrationTest.class);
+
+    private static final int NUM_RECORDS_PRODUCED = 2000;
+    private static final int NUM_TOPIC_PARTITIONS = 2;
+    private static final int CONSUME_MAX_DURATION_MS = 5000;
+
+    private EmbeddedConnectCluster connect;
+
+    @Before
+    public void setup() throws IOException {
+        connect = new EmbeddedConnectCluster(ConnectIntegrationTest.class);
+        connect.start();
+    }
+
+    @After
+    public void close() {
+        connect.stop();
+    }
+
+    /**
+     * Simple test case to bring up an embedded Connect cluster along a backing Kafka and Zk process.
+     * The test will produce and consume records, and start up a sink connector which will consume these records.
+     */
+    @Test
+    public void testProduceConsumeConnector() throws Exception {
```

Comment:
@ewencp yes, agree that these tests should not be expensive. But when you say they must be comprehensive, does that mean creating multiple tests with different converters and transformation. I don't think that should be the purpose here. We need to ensure that the interfaces of these different components work well together. Plugging in different implementations of converters ensures their correctness, but that can be tested with unit tests. Here we want to make sure that all the components (through their interfaces, work correctly with each other). For example, a test could be extended to add a transformation and/or a dead letter queue to check if all of these pieces work well together (which is what `DeadLetterQueueIntegrationTest` does, btw). Variations of this test would be running the same test (move data through a connector pipeline) in the absence of the optional components and asserting the correct behavior. 

in the light of the above comment, what about calling this test as `EndToEndSinkConnectorIntegrationTest`, and moving the test from `DeadLetterQueueIntegration` into this class as a `testWithDeadLetterQueue` test method?

@rhauch @ewencp let me know what you think?

Reply:
By comprehensive, I actually just meant that the integration test should evaluate things more end-to-end, i.e. that the integration of multiple components are working.

I agree that you are frequently testing interfaces, but also lots of frequently unspoken assumptions (or not enforced), e.g. even simple things like if `null` is a valid value for a particular argument or not, and whether other components calling the method will deliver them.

You can't cover the entire space of combinations, but I called out naming because I think the name says a lot about what you think you're testing. `DeadLetterQueueIntegrationTest` is named to be quite specifically targeted -- it's testing a single config/policy/*mostly* one class (`DeadLetterQueueReporter`).

The suggestion re: individual test methods is interesting as well -- this was also what I was getting at wrt this style of test being quite heavyweight and therefore better to keep broader. If it was an integration test with fast mocks (e.g. even if full Connect worker, mocked out ZK + broker), I'd be more inclined to have tests like `ErrorHandlingIntegrationTest` with separate tests for each. But given how expensive they are, why *not* just have a single test method in an `ErrorHandlingIntegrationTest` validate all the error handling cases end-to-end?

I guess high level what I'm getting at is that there's a tendency to write integration and system tests that validate individual features that are frequently effectively a single class + a couple of surrounding ones, but pull in a ton of other test surface area. I think getting the balance is especially hard with integration tests, but with system tests I've noticed some similar tests as well. But if you go that route, you end up with basically as many integration tests as you have unit tests, which shouldn't really be the case (the thing the integration test is testing should be larger than the unit test)?

wrt "testing the same thing", I think an end-to-end source connector and sink connector test where we swap out components in a parameterization that reasonably covers the space *is* a good set of integration tests (since it validates the source and sink data pipelines, integrated together, under at least a few different variants, e.g. different types of converter, transforms, etc). Some of those will ultimately land with the individual connectors as well (e.g. is the connector & upstream converters/transforms actually obeying the assumptions of interfaces that may not be explicit yet).

Another way to view this is that the javadoc should say *what is under test* and maybe even *why* we need this in addition to unit tests; currently this test's javadoc just says *what it will do*.